tsdb/importer/importer.go

```diff
@@ -0,0 +1,421 @@
+	if err != nil {
+		return err
+	}
+
+	expandedBlockMetas := make(map[int][]*newBlockMeta)
+	for blockIdx, meta := range blockMetas {
+		bmetas := make([]*newBlockMeta, 0)
+		if blockIdx == 0 || blockIdx == len(blockMetas)-1 {
+			// Divvy blocks outside of DB time limits into 2h blocks.
+			tpairs := makeRange(meta.mint, meta.maxt, tsdb.DefaultBlockDuration)
+			for _, tp := range tpairs {
+				m := &newBlockMeta{
+					index:     blockIdx,
+					count:     0,
+					mint:      tp.start,
+					maxt:      tp.end,
+					isAligned: false,
+					dirs:      nil,
+				}
+				bmetas = append(bmetas, m)
+			}
+		} else {
+			// Blocks that need to be aligned are kept as is.
+			bmetas = append(bmetas, meta)
+		}
+		expandedBlockMetas[blockIdx] = bmetas
+	}
+
+	err = writeSamples(bytes, tmpDbDir, sampleIndexes, expandedBlockMetas, maxSamplesInMemory, logger)
+	if err != nil {
+		return err
+	}
+
+	// Compact blocks that need to be compacted outside of DB time limits.
+	for _, metas := range expandedBlockMetas {
+		for _, meta := range metas {
+			if len(meta.dirs) > 1 {
+				compactedPath, err := compactBlocks(tmpDbDir, meta.dirs, logger)
+				if err != nil {
+					return err
+				}
+				// Delete blocks after compaction.
+				for _, dir := range meta.dirs {
+					err = os.RemoveAll(dir)
+					if err != nil {
+						return err
+					}
+				}
+				meta.dirs = []string{compactedPath}
+			}
+		}
+	}
+
+	err = fileutil.CopyDirs(tmpDbDir, dbPath)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// collectSampleInformation reads every sample, assigns each of them to a block, and collects per-block
+// information required for the actual write stage.
+func collectSampleInformation(b []byte, dbMint, dbMaxt int64, blockTimes []blockTimestampPair, logger log.Logger) ([]int, map[int]*newBlockMeta, error) {
+	var err error
+	count := 0
+	blockIndexes := make([]int, 0)
+	blockMetas := make(map[int]*newBlockMeta)
+
+	// We have at least 2 block metas - before dbMint, and after dbMaxt.
+	blockMetas[0] = &newBlockMeta{index: 0, count: 0, mint: math.MaxInt64, maxt: dbMint, isAligned: false}
+	for idx, block := range blockTimes {
+		blockMetas[idx+1] = &newBlockMeta{index: idx + 1, count: 0, mint: block.start, maxt: block.end, isAligned: true}
+	}
+	blockMetas[len(blockTimes)+1] = &newBlockMeta{index: len(blockTimes) + 1, count: 0, mint: dbMaxt, maxt: math.MinInt64, isAligned: false}
+
+	parser := textparse.New(b, contentType)
+	for {
+		var ent textparse.Entry
+		if ent, err = parser.Next(); err != nil {
+			if err == io.EOF {
+				break
+			}
+			return blockIndexes, blockMetas, err
+		}
+		if ent != textparse.EntrySeries {
+			continue
+		}
+		count += 1
+		_, ctMicroS, _ := parser.Series()
+		var ctMs int64
+		if ctMicroS == nil {
+			// Throw an error if no timestamp found.
+			return blockIndexes, blockMetas, NoTimestampError
+		} else {
+			// Open Metrics reads time as microseconds - we want milliseconds.
```

Comment:
OpenMetrics works in seconds.

Reply:
Hmm, it just seems to multiply every timestamp it sees by 1000. Which is why I have to make that change. I guess I should update the comment.